### PR TITLE
Delay skill rust for 24h after practicing

### LINKED ISSUE #51647: Excessive and unhelpful skill rust messages
### Describe the bug

Skill rust messages occur immediately when sleeping after training a skill up to a new level, if you end up exactly at the new level

### Steps To Reproduce

1. Make a new survivor, craft cudgels or other level 0 crafting task until you hit precisely level 1
2. Sleep
3. Note that you get a message about your skill decreasing

### Expected behavior

You will only be informed about skills rusting when they have meaningfully decreased from a previous maximum and not just temporarily dipped to a lower level because you went from x(0%) to x-1(99%)

Alternatively: perhaps skills shouldn't rust at all if you gained a new level in that skill within the previous 24/48/other number of hours to represent freshly learned skills sticking in your mind for a while?

### Screenshots

_No response_

### Versions and configuration

- OS: Windows
    - OS Version: MINGW/CYGWIN/MSYS2 on unknown Windows version
- Game Version: d502276 [64-bit]
- Graphics Version: Tiles
- Game Language: System language []
- Mods loaded: [
    Dark Days Ahead [dda],
    Disable NPC Needs [no_npc_food],
    Bionic Professions [package_bionic_professions]
]

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by dseguin**
#### Summary
None

#### Purpose of change
Fixes #51647

#### Describe the solution
If `SkillLevel::_lastPracticed` less that 24 hours, don't calculate skill rust. This effectively delays skill rust until the next "tick", 24 hours after practicing the skill.

#### Describe alternatives you've considered
Previously, I was using a "grace point" system to delay skill rust. The issue with that was pointed out in the comments below.

#### Testing
1. Practice carving to level up fabrication to 1
2. Wait until the 24-hour cycle recalculates skill rust
3. Notice that it now takes 2 cycles to start applying skill rust to fabrication

#### Additional context
See the previous edit of this PR to read the original description.

### CODE REVIEW

File: src/skill.cpp [hunk 1]
```diff
@@ -257,15 +257,30 @@ void SkillLevel::train( int amount, float catchup_modifier, float knowledge_modi
...
 
+    // After obtaining a new level, add rust grace points to simulate a grace period
+    if( leveled_up ) {
+        // Formula is basically the same as in SkillLevel::rust()
+        // This adds a little over one cycle's worth of rust grace points (currently 24h)
+        const int lvl_mult = ( _level + 1 ) * ( _level + 1 );
+        float lvl_exp = lvl_mult * 10000.0f;
+        float rust_div = 0.04f;
+        if( _rustAccumulator < lvl_exp * 3 ) {
+            rust_div = std::max( static_cast<float>( std::sqrt( _rustAccumulator / lvl_exp ) ), 0.04f );
```
Comment (BrettDong) [inline src/skill.cpp:279]: Is it possible that `_rustAccumulator < 0` at this moment?

Comment (dseguin) [reply src/skill.cpp:279]: I don't see why it would, but I'll add a condition just in case

Comment (BrettDong) [reply src/skill.cpp:279]: I think the NaN error is because a negative `_rustAccumulator` was passed in.

Comment (dseguin) [reply src/skill.cpp:283]: No problem, I moved the negative check to before this calculation. Hopefully that fixes it.

**Comment by I-am-Erk** [comment]
It doesn't have to be in this pr, but it would be nice if some aspect of this were moderated by a global variable, allowing mods and individual settings to easily tweak it.

**Review by I-am-Erk — APPROVED** [review]

**Review by kevingranade — CHANGES_REQUESTED** [review]
This has a lot more moving parts than just refraining from reporting on a practical skill shortfall until the magnitude is larger.

**Comment by I-am-Erk** [comment]
That's true, but I think it's reasonable to postpone rusting a newly gained skill level since it leads to some counterintuitive behaviors. Is there something specific about this system you dislike?

**Comment by I-am-Erk** [comment]
When I redid rust I skipped tracking "last practiced" because it was obnoxious but yes, that would be a better way to do this. What we could do is have "grace period" get reset to the value outlined in this PR when you practice rather than when you gain a level. That would still accomplish the goal of not dropping a newly gained level almost immediately, while not artificially enshrining level gain as a special case.

I also agree that the rust messages could be improved or simply turned off.

**Comment by dseguin** [comment]
@I-am-Erk I implemented your suggestions to make this a bit simpler. I scrapped the "grace point" stuff in favor of `SkillLevel::_lastPracticed`. I also added a menu option in the debug category to tweak it more easily.

File: src/skill.cpp [hunk 2]
```diff
@@ -320,6 +321,12 @@ bool SkillLevel::isRusty() const
            _knowledgeExperience - _exercise >= ( _level + 1 ) * ( _level + 1 ) * 10;
 }
 
+bool SkillLevel::rust_grace_period() const
```
Comment (eltank) [inline src/skill.cpp:324]: Does this need a separate function? Seems that you could simply put this code near the top of rust() below.
The name of the function also doesn't match the return value (it doesn't return a "period").

Comment (dseguin) [reply src/skill.cpp]: That's fair. I removed this function.

**Comment by dseguin** [comment]
Superseded by #51724

**Comment by dseguin** [comment]
Updated to remove menu option. The only thing this PR does is delay skill rust by 24 hours after practicing.

**Pull request merged**